README.md
# Pymoku

A Python library for the command, control and monitoring of the [Liquid Instruments Moku:Lab](http://www.liquidinstruments.com).
API documentation can be found at [ReadTheDocs](http://pymoku.readthedocs.org).

:warning: pymoku supports Moku:Lab firmware versions upto and including 511.  For more recent firmware, as well as Moku:Go and Moku:Pro support please visit https://www.liquidinstruments.com/products/apis/python-api/

# Getting Started

### 1. Requirements
- [Python](https://www.python.org) installed.  We support Python **2.7** and **3.6**.
- pip installed (`pip` for Python 2.7 or `pip3` for Python 3)
- Bonjour services installed. See [Installing Bonjour Libraries](#installing-bonjour-libraries).
- Your Moku:Lab connected to the same network as your computer.
- Internet access.

### 1. Install Pymoku
Open a command-line terminal and type

    pip install --upgrade pymoku
    moku update fetch

### 2. Update your Moku:Lab
Update your Moku:Lab with the latest firmware by typing the following in your terminal

    moku --serial=123456 update install

The serial number is the middle six digits found on the underside of the Moku:Lab. For example: XXX-123456-X.

You can also update the Moku:Lab using its IP address with

    moku --ip=192.168.0.1 update install

The IP address of your Moku:Lab device can be found with

    moku list

**NOTE:** The update process is indicated by alternating orange and white lights, or a pulsing orange light. This process can take up to 30 minutes.
### 3. Start scripting
You are now ready to control your Moku:Lab using Python! You can find a few example scripts in the **examples/** folder.
Here is a basic example of how to connect to a Moku:Lab, deploy the Oscilloscope and fetch a single hi-res data trace. Open python and run the following code

```python
from pymoku import Moku
from pymoku.instruments import Oscilloscope

m = Moku.get_by_serial(123456)
i = m.deploy_instrument(Oscilloscope)

i.set_timebase(-1e-3, 1e-3) # 1ms before and after trigger event
i.set_trigger('in1', 'rising', 0) # rising edge through 0 volts

data = i.get_data()
print(data.ch1, data.ch2, data.time)

m.close()
```

# Troubleshooting

#### moku: command not found
Ensure pymoku has been successfully installed in your Python distrubution by open a python shell and running

    import pymoku

No error indicates a successful install.

You may need to add python binaries to your PATH.  This varies with operating system and python version but as an example

    export PATH=$PATH:/home/user/.local/bin

#### Installing Bonjour Libraries
To automatically discover Moku:Lab on your network (i.e. by *name* or *serial*) you must have Bonjour installed.
Without Bonjour, your Moku:Lab will still be accessible by IP address, you just won't be able to automatically connect by name or serial number, or find it using `moku list`.
To install Bonjour:
- **Windows** install the [Bonjour Printer Services](https://support.apple.com/kb/DL999). Note that you can choose not to install this service, and
- **Linux** install the dnssd compatibility libraries. For Ubuntu, this is

        $ sudo apt-get install libavahi-compat-libdnssd1
- **OSX** comes with Bonjour by default.

#### I can't find my Moku:Lab on the network
Search your local network for running Moku:Labs with

    $ moku list

If you do not see your Moku:Lab, confirm that your Moku:Lab is powered ON, and that you have Bonjour installed on your operating system [Installing Bonjour Libraries](#installing-bonjour-libraries).

If you are connecting over
- **Access Point Mode** Ensure your computer is connected to the Moku:Lab's WiFi network.  The Moku:Lab will default to the IP address `192.168.73.1`.
- **Ethernet** Ensure the cable is plugged in an you see the orange activity light.  You will need a DHCP server running on your network to provide the Moku:Lab an IP address.
- **WiFi** After configuring the WiFi in the Moku:Lab (currently requires an iPad), you should see a solid blue light on the front.  If it's blinking indefinitely or off, check the configuration again using an iPad.
- **USB** ensure you have plugged the Moku:Lab's micro-USB port into the USB port of your PC. For Linux users, you will have to change the USB-Ethernet network settings to "Link Local" for IPv4 and IPv6. **NOTE:** Currently only one Moku:Lab can be connected over USB at a time.

See [Connecting to your Moku:Lab](https://github.com/liquidinstruments/pymoku/wiki/Connecting-to-your-Moku:Lab) guide on how to find your device's IP address.
If you are still having difficulty, contact support@liquidinstruments.com.

#### ImportError: No Module named pymoku
Make sure you are running the version of Python you installed pymoku to.  Often a system will have multiple Python installations. Try substituting `pip` with `python -m pip` in the installation.
If you installed pymoku inside an Environment (i.e. via virtualenv or conda-env), ensure that Environment is activated. You can check that pymoku is installed in your currently running environment using

    (myenv)$ pip list

or

    (myenv)$ conda list # For conda users

#### zmq.error.Again: Resource temporarily unavailable
1. Ensure your Moku:Lab is visible on the network by calling `moku list` or searching for the service using BonjourBrowser
2. Make sure you are using the latest version of pymoku, and that your Moku:Lab has been upgraded to the newest firmware. See the Installation section above for details.
#### pymoku.DeployException: Deploy Error 3
Confirm you have have the latest binary data pack and that your Moku:Lab is running the latest firmware

    $ moku update fetch
    $ moku --serial=123456 update install

## Issue Tracking

Please log issues here on Github.


CHANGELOG.md
# Release Notes

## 2.7.0
- Renamed BodeAnalyzer to FrequencyResponseAnalyzer.
- FrequencyResponseAnalyzer
    - can set phase difference between outputs
    - can demodulate at up to 15th harmonic
- minor bug fixes

## 2.6.0
- Update for Moku:Lab firmware version 1.8 (501)
- minor bug fixes

## 2.3.0
### Headlines
- Added **new** instrument support for the **FIR Filter Box**:
	- Build FIR filters with sample rates of up to 15 MS/s.
	- Create arbitrary impulse responses by loading up to 14,000 custom coefficients.
	- 2 input channels, 2 output channels with control matrix for blending inputs.
- Added advanced triggering option for the Oscilloscope:
	- Max/min pulse width
- Added advanced output triggering options to the Arbitrary Waveform Generator, including:
	- Trigger type (edge, pulse-width)
	- Trigger source (ADCs, external)
	- Triggered output modes (single, continuous)
	- Triggered output duration
- Support for data logging instruments to log directly to MATLAB (`.mat`) and NumPy (`.npy`) file types
- Simplified pip pymoku installation process
	- Removed various library dependencies
	- `moku_convert` command-line conversion tool falls back to slower Python implementation without `liquidreader` C-library installed.
- Network packets are now encrypted
- Latest Moku:Lab instrument files and firmware can now be downloaded and installed via the `moku` command line tool for easier device updating, e.g:
	- `moku update fetch`
	- `moku --serial=123456 update install`

### API Changes
- `Moku.deploy_or_connect(...)` and `Moku.deploy_instrument(...)` now have matching API (i.e., input parameters and return value). **Note:** this does not imply identical behaviour (see API documentation).
- `IIRFilterBox` instrument class functions have been changed to match `FIRFilter` API.
	- `set_offset_gain` renamed to `set_gains_offsets` and now only configures the input/output offsets and gains of a specified filter channel.
	- `set_control_matrix` function added to configure input signal mixing gain coefficients.
	- Updated various input parameter ranges including +-1.0V input offset, and +-2.0V output offset.
- Class documentation of various instruments updated.

### Bug Fixes
- IIR Filter outputs are scaled correctly
- Waveform Generator DAC trigger thresholds set correctly
- Oscilloscope and Datalogger instruments correctly log Channel 2 to CSV
- Python 3 fixes:
	- `Moku.list_mokus()` runs without error

### Known Issues
- `LockInAmp.set_outputs` does not change output signal for all selections.

## 2.2.1
### Bug Fixes

- All discovery protocols now work for Python 3
- Moku:PID Differentiator saturation now works correctly when the double-integrator is enabled
- Manually-installable pip package was missing build files

## 2.2.0
### Headlines

- Full support for all instruments, including for the first time:
	- Bode Analyser
    - Filter Box
    - Arbitrary Waveform Generator
    - PID Controller
- New phase-locked output for Phasemeter
- New triggering on Waveform Generator

### API Changes

- All objects and functions now follow US Spelling conventions; e.g. SpectrumAnalyser → SpectrumAnalyzer.

### Known Issues

- The Moku.list_mokus() discovery function doesn't work on Python 3
- The PID Controller differentiator saturation and double-integrator have an unexpected (buggy) interaction, it's recommended to only use on of these features at a time.

## 2.1.3
### Bug Fixes

- Fixed Oscilloscope embedded waveform generator instance reference

## 2.1.2
### Bug Fixes

- Python 3 syntax error

## 2.1.1
### Bug Fixes

- Documentation updates only

## 2.1.0
### Headlines

- New support for:
	- Lock-in Amplifier
	- Bode Analyser
- Numerous bug fixes

### API Changes
- Signal Generator renamed Waveform Generator
	- This is for consistency with documentation and other releases.  The class name has changed, please update your scripts.
- Oscilloscope triggering hysteresis cannot be explicitly set in volts.
	- Complexities with the DSP meant this was always buggy. Hysteresis can now just be turned on and off, toggled between values that are used in the iPad and other clients to implement noise rejection.
- Removed 'upload' parameter from start_data_log.
	- This parameter never worked well, the user should call the instrument's upload function after log completion.
- The meaning of Phasemeter rate flags has been re-defined.
	- 'fast' and 'slow' have been replaced with 'fast', 'medium' and 'slow', representing approximately 1.9ksps, 120sps and 30sps output rate respectively.
### Bug Fixes

- Hysteresis voltage was not correctly set
- Spectrum Analyser gain corrections were slightly wrong
- Data logger now enforces a minimum sample rate rather than just crashing on low rates
- pymoku couldn't be installed on Windows through pip
- autocommit sometimes stopped working if a function through an exception
- numerous compatibility issues with Python 3 rectified
- numerous parameter bounds checks made more explicit and the error messages more helpful
- numerous documentation fixes

### Known Issues
- Doesn't support all released instruments
- Only the iPad can currently be used for PID Controller
- pymoku cannot update Moku:Lab firmware
	- In order to ensure correct functionality of your device, you should use an iPad to check that your Moku:Lab's firmware is up to date when you upgrade your pymoku version. Some pymoku 2.1 functionality may not work unless you use your iPad to update said firmware.
- pymoku Datalogger will not warn if you are about to log a file larger than the maximum file size supported by your SD Card.
	- Most SD cards have a 4GB maximum file size due to limitations of the FAT32 filesystem. If you try and log a file that will exceed that, pymoku will allow it to start but will error-out after it has reached the 4GB limit.
- Waveform Generator output voltages assume a 50-Ohm load

pymoku/tools/moku.py
#!/usr/bin/env python

from argparse import ArgumentParser
import os, os.path, shutil, tempfile, tarfile
import requests, hashlib, sys, threading

from pymoku import *
from pymoku.tools.compat import *

MOKUDATAURL = 'http://updates.liquidinstruments.com/static/' + MOKUDATAFILE

import logging
logging.basicConfig(format='%(message)s', level=logging.INFO)

parser = ArgumentParser()
subparsers = parser.add_subparsers(title="action", dest='action', description="Action to take")
subparsers.required = True

# Global arguments
parser.add_argument('--serial', default=None, help="Serial Number of the Moku to connect to")
parser.add_argument('--name', default=None, help="Name of the Moku to connect to")
parser.add_argument('--ip', default=None, help="IP Address of the Moku to connect to")
parser.add_argument('--force', action='store_true', help="Bypass compatibility checks with the target Moku. Don't touch unless you know what you're doing.")

# View and load new instrument bitstreams
def update(args):
    if args.action == 'fetch':
        url = args.url

        try:
            r = requests.get(url.replace('tar.gz', 'md5'))
            r.raise_for_status() # Checks for any HTTP errors
            remote_hash = r.text.split(' ')[0]

            with open(DATAPATH + '/' + MOKUDATAFILE, 'rb') as f:
                assert hashlib.md5(f.read()).hexdigest() == remote_hash and not args.force

        except requests.HTTPError as e:
            logging.error("ERROR: Unable to retrieve updates from server.\n%s" % e.message)
            return
        except:
            # Any non-HTTP exception causes new archive to download
            logging.info("Fetching data pack from: %s" % url)
            with requests.get(url, stream=True) as r:
                r.raise_for_status() # Check for any HTTP errors
                length = int(r.headers['content-length'])
                recvd = 0

                with open(DATAPATH + '/' + MOKUDATAFILE, 'wb') as f:
                    for chunk in r.iter_content(chunk_size=400000):
                        f.write(chunk)
                        recvd = recvd + len(chunk)
                        sys.stdout.write("\r[%-30s] %3d%%" % ('#' * int(30.0 * recvd / length), (100.0 * recvd / length)))
                        sys.stdout.flush()
                    sys.stdout.write('\r[%-30s] Done!\n' % ('#' * 30))

            with open(DATAPATH + '/' + MOKUDATAFILE, 'rb') as f:
                assert hashlib.md5(f.read()).hexdigest() == remote_hash
        else:
            logging.info("%s Already up to date" % MOKUDATAFILE)
    elif args.action == 'install':
        firmware_reboot = False
        patch_reboot = False
        moku = connect(args, force=True)

        v = moku.get_hw_version()
        f = DATAPATH + '/' + MOKUDATAFILE

        def _load_firmware():
            tardata = tarfile.open(DATAPATH + '/' + MOKUDATAFILE)
            fw_tarinfo = tardata.getmember('moku%2d.fw' % (v * 10))
            fw_file = tardata.extractfile(fw_tarinfo)
            moku._send_file_bytes('f', 'moku.fw', fw_file.read())
            fw_file.close()
            tardata.close()
            return True

        def _load_patches():
            moku._delete_packs()
            tardata = tarfile.open(DATAPATH + '/' + MOKUDATAFILE)
            tar_packs = [ f for f in tardata.getnames() if f.endswith(('hgp','hgp.aes')) ]
            for p in tar_packs:
                pack_name = p.split('/')[-1]
                pack_tarinfo = tardata.getmember(p)
                pack_file = tardata.extractfile(pack_tarinfo)
                logging.info("Installing pack - %s" % pack_name)
                moku._send_file_bytes('p', pack_name, pack_file.read())
                pack_file.close()
            tardata.close()
            return True

        old_fw = moku.get_firmware_build()
        new_fw = version.compat_fw[0]
        logging.info('Updating Moku %06d from %d to %d...' % (int(moku.get_serial()), old_fw, new_fw))

        if args.force:
            if old_fw > new_fw:
                logging.warning('Downgrading firmware.')
            firmware_reboot |= _load_firmware()
        else:
            if old_fw == new_fw:
                logging.info('Firmware already up to date.')
            elif old_fw > new_fw:
                logging.error('Refusing to downgrade firmware.')
            else:
                firmware_reboot |= _load_firmware()

        if firmware_reboot:
            serial_num = int(moku.get_serial())
            moku._trigger_fwload()
            log.info("Successfully started firmware update.\n"
                     "Your Moku:Lab will shut down automatically when complete. "
                     "This process can take up to 30 minutes.\n"
                     "Please run `moku --serial=%d update install` "
                     "a second time to ensure all updates are installed correctly." % (serial_num))
            return

        # Applying patches
        logging.info('Checking for updates.')

        if not patch_is_compatible(moku):
            logging.info('Applying updates...')
            patch_reboot |= _load_patches()
        else:
            logging.info('Moku:Lab already up to date.')

        if patch_reboot:
            moku._restart_board()
            log.info("Successfully applied updates. Your Moku:Lab will now shut down to complete the update process.")

parser_update = subparsers.add_parser('update', help="Check and update instruments on the Moku.")
parser_update.add_argument('--url', help='Override location of data pack', default=MOKUDATAURL)
parser_update.add_argument('--force', action='store_true')
parser_update.add_argument('action', help='Action to perform', choices=['fetch', 'install'])
parser_update.set_defaults(func=update)

def listmokus(args):
    mokus = pymoku.BonjourFinder().find_all(timeout=2.0)
    mokus.sort()
    print("{: <20} {: >6} {: >6} {: >15}".format('Name', 'Serial', 'FW', 'IP'))
    print("-" * (20 + 6 + 6 + 15 + 3))

    def _querytask(m):
        x = None
        try:
            x = Moku(m, force=True)
            print("{: <20} {:>06} {: 6d} {: >15}".format(
                x.get_name()[:20],
                int(x.get_serial()),
                x.get_firmware_build(),
                m))
        except:
            print("Couldn't query IP %s" % m)
        finally:
            try:
                x.close()
            except:
                pass

    tasks = []
    for m in mokus:
        tasks.append(threading.Thread(target=_querytask, args=[m]))

    for t in tasks:
        t.start()
    for t in tasks:
        t.join()

parser_list = subparsers.add_parser('list', help="List Moku:Labs on the network.")
parser_list.set_defaults(func=listmokus)

def query_property(args):
    moku = None
    try:
        moku = connect(args, force=True)
        if args.value:
            moku._set_property_single(args.property, args.value)
        print("%s = %s" % (args.property, moku._get_property_single(args.property)))
    finally:
        if moku:
            moku.close()

parser_set = subparsers.add_parser('property', help="Set Moku:Lab property.")
parser_set.add_argument('property', help='Property to query')
parser_set.add_argument('value', help='Value to write to property', nargs='?')
parser_set.set_defaults(func=query_property)

# View and load new instrument bitstreams
def instrument(args):
    moku = None
    try:
        moku = connect(args, force=True)
        if args.action == 'load':
            if not len(args.files):
                print("No instrument files specified for loading")
                return

            for file in args.files:
                fname = os.path.basename(file)
                chk = moku._load_bitstream(file)
                print("Successfully loaded new instrument {} version {}".format(fname, chk))
        else:
            exit(1)
    finally:
        if moku:
            moku.close()

parser_instruments = subparsers.add_parser('instrument', help="Check and update instruments on the Moku.")
parser_instruments.add_argument('action', help='Action to take', choices=['load'])
parser_instruments.add_argument('files', nargs='*', default=None, help="Path to local instrument file(s), if any")
parser_instruments.set_defaults(func=instrument)

# View and load new packages
def package(args):
    moku = None
    try:
        moku = connect(args)
        if args.action == 'load':
            if not args.file or not args.file.endswith(('hgp', 'hgp.aes')):
                print('Package load requires an HGP file to be specified')
                return

            fname = os.path.basename(args.file)
            moku._send_file('p', args.file)

            print("Successfully loaded new package {}".format(fname))
        else:
            exit(1)
    finally:
        if moku:
            moku.close()

parser_package = subparsers.add_parser('package', help="Check and load special feature packages on the Moku.")
parser_package.add_argument('action', help='Action to perform', choices=['load'])
parser_package.add_argument('file', nargs='?', default=None, help="Path to local package file, if any")
parser_package.set_defaults(func=package)


# View firmware version and load new versions.
def firmware(args):
    moku = None

    try:
        moku = connect(args, force=True)
        if args.action == 'version':
            print("Moku Firmware Version {}".format(moku.get_firmware_build()))
        elif args.action == 'load':
            v = moku.get_hw_version()
            if not args.file:
                f = DATAPATH + '/moku%2d.fw' % (v * 10) #TODO
            else:
                f = args.file

            if not f.endswith('fw'):
                print('Package load requires an FW file to be specified')
                return

            print("Loading firmware from: %s" % f)
            moku._load_firmware(f)
            print("Successfully started firmware update. Your Moku will shut down automatically when complete.")
        elif args.action == 'check_compat':
            compat = firmware_is_compatible(moku)

            if compat == None:
                print("Unable to determine compatibility, perhaps you're running a development or pre-release build?")
            else:
                print("Compatible." if compat else "Moku:Lab firmware {} incompatible with Pymoku v{}. "
                    "Please update using\n moku --ip={} update fetch\n moku --ip={} update install"
                    .format(moku.get_firmware_build(), PYMOKU_VERSION, moku.get_ip(), moku.get_ip()))
        elif args.action == 'restart':
            moku._restart_board()
            print("Moku rebooted.")
        else:
            exit(1)
    finally:
        try:
            if moku:
                moku.close()
        except zmq.error.Again:
            pass # Firmware update can stop us being able to close

parser_firmware = subparsers.add_parser('firmware', help="Check and load new firmware for your Moku.")
parser_firmware.add_argument('action', help='Action to perform', choices=['version', 'load', 'check_compat', 'restart'])
parser_firmware.add_argument('file', nargs='?', default=None, help="Path to local firmware file, if any")
parser_firmware.set_defaults(func=firmware)

def main():
    logging.info("PyMoku %s" % PYMOKU_VERSION)
    args = parser.parse_args()
    args.func(args)

def connect(args, force=False):
    if len([ x for x in (args.serial, args.name, args.ip) if x]) != 1:
        print("Please specify exactly one of serial, name or IP address of target Moku")
        exit(1)

    force = force or args.force
    if args.serial:
        moku = Moku.get_by_serial(args.serial, force=force)
        print(moku.get_name())
    elif args.name:
        moku = Moku.get_by_name(args.name, force=force)
    else:
        moku = Moku(args.ip, force=force)

    return moku


# Compatible with direct run and distutils binary packaging
if __name__ == '__main__':
    main()
